README
String::Sections

Data Section sports the following default data markup

  __[ somename ]__
  Data
  __[ anothername ]__
  More data

This module is designed to behave as a work-alike, except on already
extracted string data.

INSTALLATION

This is a Perl module distribution. It should be installed with whichever
tool you use to manage your installation of Perl, e.g. any of

  cpanm .
  cpan  .
  cpanp -i .

Consult http://www.cpan.org/modules/INSTALL.html for further instruction.
Should you wish to install this module manually, the procedure is

  perl Build.PL
  ./Build
  ./Build test
  ./Build install

COPYRIGHT AND LICENSE

This software is copyright (c) 2015 by Kent Fredric <kentnl@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Release history for String-Sections

0.003003 2015-04-03T23:00:46Z 3be462c

0.3.2 2013-09-26T07:43:21Z
 [00 Minor]
 - Attempt to avoid Test::Compile bugs on Win32

 [Dependencies::Added / test requires]
 - File::Spec

 [Dependencies::Changed / develop suggests]
 - Dist::Zilla::PluginBundle::Author::KENTNL v1.8.2 → v1.8.3

0.3.1 2013-08-30T20:45:10Z
 [!Minor]
 - Bump to get newer Test::Compile

 [Dependencies::Added / develop requires]
 - Test::Kwalitee 1.08

 [Dependencies::Added / test requires]
 - IO::Handle
 - IPC::Open3

 [Dependencies::Changed / develop suggests]
 - Dist::Zilla::PluginBundle::Author::KENTNL v1.8.0 → v1.8.2

 [Dependencies::Removed / test requires]
 - Capture::Tiny
 - blib

0.3.0 2013-08-11T07:11:18Z
 [!Major]
 - section_names now retains document order where possible.
 - use section_names_sorted if you need them sorted
 - More type constraints inside.

 [Dependencies::Added / runtime requires]
 - Types::Standard

 [Dependencies::Added / test requires]
 - Capture::Tiny
 - blib

 [Dependencies::Changed / build requires]
 - Module::Build 0.4005 → 0.4007

 [Dependencies::Changed / configure requires]
 - Module::Build 0.4005 → 0.4007

 [Dependencies::Changed / develop suggests]
 - Dist::Zilla::PluginBundle::Author::KENTNL v1.7.4 → v1.8.0

 [Dependencies::Removed / runtime requires]
 - Params::Classify

 [Dependencies::Removed / test requires]
 - File::Find
 - File::Temp

0.2.3 2013-07-16T03:37:29Z
 [Bug Fixes]
 - Need recent Moo for builder => sub {}

 [Dependencies::Changed / runtime requires]
 - Moo 0 → 1.000008

0.2.2 2013-06-25T04:12:32Z
 [Bug Fixes]
 - load_filehandle used ->_default name instead of ->default_name
 - _isa_string tried to call ::Classif<t> instead of ::Classif<y>
 - shallow_merge called blessed() instead of _blessed()

 [Dependencies::Added / test requires]
 - Data::Dump
 - FindBin
 - Path::Tiny
 - Sub::Exporter::Progressive
 - lib

 [Documentation]
 - default_name now stubbed

 [Tests]
 - Much more thoughrough coverage of load_file and load_list behaviour.
 - Generated tests Facilitated by an arsenal of configuration and templates, using String::Sections to generate the
   templated tests as well.
 - devel/compile_test_templates.pl is thus an example of potential usage.

0.2.1 2013-06-21T12:10:27Z
 [Dependencies::Removed / runtime requires]
 - Sub::Quote

 [Documentation]
 - Misc Documentation enhancement

0.2.0 2013-06-20T19:58:41Z
 [00_NOTEWORTHY Changes]
 - Interface majorly refactored and split.
 - Main class is now an instance factory instead of a self-mutator.

 [Dependencies::Added / develop requires]
 - Pod::Coverage::TrustPod
 - Test::CPAN::Changes 0.19
 - Test::CPAN::Meta
 - Test::Pod 1.41
 - Test::Pod::Coverage 1.08
 - version 0.9901

 [Dependencies::Added / runtime requires]
 - Params::Classify
 - strict
 - warnings

 [Dependencies::Changed / build requires]
 - Module::Build 0.3601 → 0.4005

 [Dependencies::Changed / configure requires]
 - Module::Build 0.3601 → 0.4005

 [Dependencies::Changed / develop recommends]
 - Dist::Zilla::PluginBundle::Author::KENTNL::Lite 0.01009803 → v1.3.0

 [Dependencies::Changed / develop suggests]
 - Dist::Zilla::PluginBundle::Author::KENTNL v1.0.0 → v1.7.4

 [Dependencies::Changed / test requires]
 - Test::More 0.96 → 0.98

 [Dependencies::Removed / runtime recommends]
 - Params::Classify

 [Dependencies::Removed / test requires]
 - English

 [Documentation]
 - Update copyright year.
 - Update LICENSE address
 - Added README.mkdn to project

 [Internals]
 - Add $AUTHORITY var in all packages
 - Move $VERSION var out of BEGIN {}

 [Meta]
 - adds x_authority key
 - bug tracker to github issues
 - git URIs moved to https:// protocols

 [Packaging]
 - utilises Module::Build 0.4004 test_requires
 - author/release tests no longer in t/

0.1.1 2011-06-07T16:22:34Z
 [Dependencies]
 - No longer uses Package::Stash
 - Adds Moo and Sub::Quote ( part of Moo )
 - No longer recommends Sub::Name for develop time.

 [Documentation]
 - Eliminated Documentation referring to experiemental ideas that are now righly removed.

 [Features]
 - Implemented load_filehandle

 [Internals]
 - Refactored Line adding logic to its own private function.
 - Eliminated all the hand-rolled OO in favour of Moo.
 - Reworked many internal things to rely on Moo code instead.
 - Eliminated lots of redundant foolish lazyload code.

0.1.0 2011-04-30T02:11:31Z
 - First version.



t/lib/Test/Fatal/Assert.pm
use strict;
use warnings;

package Test::Fatal::Assert;
use Test::More;
use Test::Fatal qw( exception );
use Sub::Exporter::Progressive -setup => {
  exports => [qw( nofatals fatals and_fatal_is )],
  groups  => {
    default => [qw( nofatals fatals and_fatal_is )],
  },
};

our $REPORT_PASS;

sub nofatals($$;$) {
  my ( $reason, $code, $callback ) = @_;
  my $result = exception { $code->() };
  if ($result) {
    fail("$reason raised no exception");
    diag($result);
    if ($callback) {
      local $_ = $result;
      $callback->($result);
    }
  }
  else {
    $REPORT_PASS and pass("$reason raised no exception");
  }
}

sub fatals($$;$) {
  my ( $reason, $code, $callback ) = @_;
  my $result = exception { $code->() };
  if ( not $result ) {
    fail("$reason should raise exception");
  }
  else {
    pass("$reason should raise exception");
    if ( defined $callback ) {
      local $_ = $result;
      $callback->($result);
    }
  }
}

sub and_fatal_is(&) {
  my ($code) = @_;
  return $code;
}
1;


lib/String/Sections.pm
use strict;
use warnings;

package String::Sections;
$String::Sections::VERSION = '0.003003';
# ABSTRACT: Extract labeled groups of sub-strings from a string.













































use 5.010001;
use Moo;
use Types::Standard qw( RegexpRef Str Bool Maybe FileHandle ArrayRef );

our $TYPE_REGEXP          = RegexpRef;
our $TYPE_OPTIONAL_STRING = Maybe [Str];
our $TYPE_BOOL            = Bool;
our $TYPE_FILEHANDLE      = FileHandle;
our $TYPE_INPUT_LIST      = ArrayRef [Str];



















































































sub _croak { require Carp; goto &Carp::croak; }





sub __add_line {

  my ( $self, $stash, $line ) = @_;

  if ( ${$line} =~ $self->header_regex ) {
    $stash->set_current("$1");
    $stash->append_data_to_current_section();
    return 1;    # 1 is next.
  }

  # not valid for lists because lists are not likely to have __END__ in them.
  if ( $self->stop_at_end ) {
    return 0 if ${$line} =~ $self->document_end_regex;
  }

  if ( $self->ignore_empty_prelude ) {
    return 1 if not $stash->has_current and ${$line} =~ $self->empty_line_regex;
  }

  if ( not $stash->has_current ) {
    _croak( 'bogus data section: text outside named section. line: ' . ${$line} );
  }

  if ( $self->enable_escapes ) {
    my $regex = $self->line_escape_regex;
    ${$line} =~ s{$regex}{}msx;
  }

  $stash->append_data_to_current_section($line);

  return 1;
}



























sub load_list {
  my ( $self, @rest ) = @_;
  $TYPE_INPUT_LIST->assert_valid( \@rest );
  require String::Sections::Result;
  my $result_ob = String::Sections::Result->new();

  if ( $self->default_name ) {
    $result_ob->set_current( $self->default_name );
    $result_ob->append_data_to_current_section();
  }

  for my $line (@rest) {
    my $result = $self->__add_line( $result_ob, \$line );
    next if $result;
    last if not $result;

  }
  return $result_ob;
}







sub load_string {
  my ( $self, $string ) = @_;
  return _croak('Not Implemented');
}







sub load_filehandle {
  my ( $self, $fh ) = @_;

  $TYPE_FILEHANDLE->assert_valid($fh);
  require String::Sections::Result;
  my $result_ob = String::Sections::Result->new();

  if ( $self->default_name ) {
    $result_ob->set_current( $self->default_name );
    $result_ob->append_data_to_current_section();
  }
  while ( defined( my $line = <$fh> ) ) {
    my $result = $self->__add_line( $result_ob, \$line, );
    next if $result;
    last if not $result;
  }

  return $result_ob;

}

#
# Defines accessors *_regex and _*_regex , that are for public and private access respectively.
# Defaults to _default_*_regex.
#





sub _regex_type {
  my $name = shift;
  return ( is => 'ro', isa => $TYPE_REGEXP, builder => '_default_' . $name, lazy => 1 );
}





sub _string_type {
  my $name = shift;
  return ( is => 'ro', isa => $TYPE_OPTIONAL_STRING, builder => '_default_' . $name, lazy => 1 );
}





sub _boolean_type {
  my $name = shift;
  return ( is => 'ro', isa => $TYPE_BOOL, builder => '_default_' . $name, lazy => 1 );
}







has 'header_regex' => _regex_type('header_regex');







has 'empty_line_regex' => _regex_type('empty_line_regex');







has 'document_end_regex' => _regex_type('document_end_regex');







has 'line_escape_regex' => _regex_type('line_escape_regex');







has 'default_name' => _string_type('default_name');

# Boolean Accessors







has 'stop_at_end' => _boolean_type('stop_at_end');







has 'ignore_empty_prelude' => _boolean_type('ignore_empty_prelude');







has 'enable_escapes' => _boolean_type('enable_escapes');

# Default values for various attributes.





sub _default_header_regex {
  return qr{
    \A                # start
      _+\[            # __[
        \s*           # any whitespace
          ([^\]]+?)   # this is the actual name of the section
        \s*           # any whitespace
      \]_+            # ]__
      [\x0d\x0a]{1,2} # possible cariage return for windows files
    \z                # end
    }msx;
}





sub _default_empty_line_regex {
  return qr{
      ^
      \s*   # Any empty lines before the first section get ignored.
      $
  }msx;
}





sub _default_document_end_regex {
  return qr{
    ^          # Start of line
    __END__    # Document END matcher
  }msx;
}





sub _default_line_escape_regex {
  return qr{
    \A  # Start of line
    \\  # A literal \
  }msx;
}





sub _default_default_name { return }





sub _default_stop_at_end { return }





sub _default_ignore_empty_prelude { return 1 }





sub _default_enable_escapes { return }

1;

__END__

=pod

=head1 NAME

String::Sections - Extract labeled groups of sub-strings from a string.

=head1 VERSION

version 0.003003

=head1 SYNOPSIS

  use String::Sections;

  my $sections = String::Sections->new();
  my $result = $sections->load_list( @lines );
  # TODO
  # $sections->load_string( $string );
  # $sections->load_filehandle( $fh );
  #
  # $sections->merge( $other_sections_object );

  my @section_names = $result->section_names();
  if ( $result->has_section( 'section_label' ) ) {
    my $string_ref = $result->section( 'section_label' );
    ...
  }

=head1 DESCRIPTION

Data Section sports the following default data markup

  __[ somename ]__
  Data
  __[ anothername ]__
  More data

This module is designed to behave as a work-alike, except on already extracted string data.

=head1 METHODS

=head2 C<new>

=head2 C<new( %args )>

  my $object = String::Sections->new();

  my $object = String::Sections->new( attribute_name => 'value' );

=head2 C<load_list ( @strings )>

  my @strings = <$fh>;

  my $result = $string_section->load_list( @strings );

This method handles data as if it had been slopped in unchomped from a filehandle.

Ideally, each entry in @strings will be terminated with $/ , as the collated data from each section
is concatenated into a large singular string, e.g.:

  $result = $string_section->load_list("__[ Foo ]__\n", "bar\n", "baz\n" );
  $section_foo = $result->section('Foo')
  # bar
  # baz

  $result = $s_s->load_list("__[ Foo ]__\n", "bar", "baz" );
  $result->section('Foo');
  # barbaz

  $object->load_list("__[ Foo ]__", "bar", "baz" ) # will not work by default.

This behaviour may change in the future, but this is how it is with the least effort for now.

=head2 C<load_string>

TODO

=head2 C<load_filehandle( $fh )>

  my $result = $object->load_filehandle( $fh )

=head2 C<header_regex>

=head2 C<empty_line_regex>

=head2 C<document_end_regex>

=head2 C<line_escape_regex>

=head2 C<default_name>

=head2 C<stop_at_end>

=head2 C<ignore_empty_prelude>

=head2 C<enable_escapes>

=head1 ATTRIBUTES

=head2 C<header_regex>

=head2 C<empty_line_regex>

=head2 C<document_end_regex>

=head2 C<line_escape_regex>

=head2 C<default_name>

=head2 C<stop_at_end>

=head2 C<ignore_empty_prelude>

=head2 C<enable_escapes>

=head1 PRIVATE METHODS

=head2 C<__add_line>

=head2 C<_default_header_regex>

=head2 C<_default_empty_line_regex>

=head2 C<_default_document_end_regex>

=head2 C<_default_line_escape_regex>

=head2 C<_default_default_name>

=head2 C<_default_stop_at_end>

=head2 C<_default_ignore_empty_prelude>

=head2 C<_default_enable_escapes>

=head1 PRIVATE FUNCTIONS

=head2 C<_croak>

=head2 C<_regex_type>

=head2 C<_string_type>

=head2 C<_boolean_type>

=begin MetaPOD::JSON v1.1.0

{
    "namespace":"String::Sections",
    "interface":"class",
    "inherits":"Moo::Object"
}


=end MetaPOD::JSON

=head1 DEVELOPMENT

This code is still new and under development.

All the below facets are likely to change at some point, but don't
largely contribute to the API or usage of this module.

=over 4

=item * Needs Perl 5.10.1

To make some of the development features easier.

=back

=head1 Recommended Use with Data::Handle

This modules primary inspiration is Data::Section, but intending to split and decouple many of the
internal parts to add leverage to the various behaviors it contains.

Data::Handle solves part of a problem with Perl by providing a more reliable interface to the __DATA__ section in a file that is not impeded by various things that occur if its attempted to be read more than once.

In future, I plan on this being the syntax for connecting Data::Handle with this module to emulate Data::Section:

  my $dh = Data::Handle->new( __PACKAGE__ );
  my $ss = String::Sections->new( stop_at_end => 1 );
  my $result = $ss->load_filehandle( $dh );

This doesn't implicitly perform any of the inheritance tree magic Data::Section does,
but its also planned on making that easy to do when you want it with C<< ->merge( $section ) >>

For now, the recommended code is not so different:

  my $dh = Data::Handle->new( __PACKAGE__ );
  my $ss = String::Sections->new( stop_at_end => 1 );
  my $result  = $ss->load_list( <$dh> );

Its just somewhat less efficient.

=head1 SIGNIFICANT CHANGES

=head2 Since 0.1.x

=head3 API

In 0.1.x, C<API> was

    my $section = String::Sections->new();
    $section->load_*( $source );
    $section->section_names

This was inherently fragile, and allowed weird things to occur when people
tried to get data from it without it being populated yet.

So starting with 0.2.0, the C<API> is

    my $section = String::Sections->new();
    my $result  = $section->load_*( $source );
    $result->section_names;

And the main class is a factory for L<< C<String::Sections::Result>|String::Sections::Result >> objects.

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2015 by Kent Fredric <kentnl@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/String/Sections/Result.pm
use strict;
use warnings;

package String::Sections::Result;
$String::Sections::Result::VERSION = '0.003003';
# ABSTRACT: Glorified wrapper around a hash representing a parsed String::Sections result
#













use Moo 1.000008;

## no critic (RequireArgUnpacking)

sub _croak   { require Carp;         goto &Carp::croak; }
sub _blessed { require Scalar::Util; goto &Scalar::Util::blessed }














use Types::Standard qw( HashRef ArrayRef ScalarRef Str Maybe );

our $TYPE_SECTION_NAME     = Str;
our $TYPE_SECTION_NAMES    = ArrayRef [Str];
our $TYPE_SECTION          = ScalarRef [Str];
our $TYPE_OPTIONAL_SECTION = Maybe [$TYPE_SECTION];
our $TYPE_CURRENT          = Maybe [Str];
our $TYPE_SECTIONS         = HashRef [$TYPE_SECTION];

has 'sections' => (
  is      => ro =>,
  isa     => $TYPE_SECTIONS,
  lazy    => 1,
  builder => sub {
    return {};
  },
);




















has '_current' => (
  is        => ro  =>,
  isa       => $TYPE_CURRENT,
  reader    => '_current',
  writer    => 'set_current',
  predicate => 'has_current',
  lazy      => 1,
  builder   => sub { return _croak('current never set, but tried to use it') },
);







has '_section_names' => (
  is   => ro =>,
  isa  => $TYPE_SECTION_NAMES,
  lazy => 1,
  builder => sub { return [] },
);








sub section {
  $TYPE_SECTION_NAME->assert_valid( $_[1] );
  return $_[0]->sections->{ $_[1] };
}









sub section_names { return ( my @list = @{ $_[0]->_section_names } ) }





sub section_names_sorted { return ( my @list = sort @{ $_[0]->_section_names } ) }









sub has_section {
  $TYPE_SECTION_NAME->assert_valid( $_[1] );
  return exists $_[0]->sections->{ $_[1] };
}







sub set_section {
  $TYPE_SECTION_NAME->assert_valid( $_[1] );
  $TYPE_SECTION->assert_valid( $_[2] );
  if ( not exists $_[0]->sections->{ $_[1] } ) {
    push @{ $_[0]->_section_names }, $_[1];
  }
  $_[0]->sections->{ $_[1] } = $_[2];
  return;
}










sub append_data_to_current_section {
  $TYPE_OPTIONAL_SECTION->assert_valid( $_[1] );
  if ( not exists $_[0]->sections->{ $_[0]->_current } ) {
    push @{ $_[0]->_section_names }, $_[0]->_current;
    my $blank = q{};
    $_[0]->sections->{ $_[0]->_current } = \$blank;
  }
  if ( defined $_[1] ) {
    ${ $_[0]->sections->{ $_[0]->_current } } .= ${ $_[1] };
  }
  return;
}










sub append_data_to_section {
  $TYPE_SECTION_NAME->assert_valid( $_[1] );
  $TYPE_OPTIONAL_SECTION->assert_valid( $_[2] );
  if ( not exists $_[0]->sections->{ $_[1] } ) {
    push @{ $_[0]->_section_names }, $_[1];
    my $blank = q{};
    $_[0]->sections->{ $_[1] } = \$blank;
  }
  if ( defined $_[2] ) {
    ${ $_[0]->sections->{ $_[1] } } .= ${ $_[2] };
  }
  return;
}











sub shallow_clone {
  my $class = _blessed( $_[0] ) || $_[0];
  my $instance = $class->new();
  for my $name ( $_[0]->section_names ) {
    $instance->set_section( $name, $_[0]->sections->{$name} );
  }
  return $instance;
}














sub shallow_merge {
  my $class = _blessed( $_[0] ) || $_[0];
  my $instance = $class->new();
  for my $name ( $_[0]->section_names ) {
    $instance->set_section( $name, $_[0]->sections->{$name} );
  }
  for my $name ( $_[1]->section_names ) {
    $instance->set_section( $name, $_[1]->sections->{$name} );
  }
  return $instance;
}







sub _compose_section {
  $TYPE_SECTION_NAME->assert_valid( $_[1] );
  return sprintf qq[__[%s]__\n%s], $_[1], ${ $_[0]->sections->{ $_[1] } };
}







sub to_s {
  my $self = $_[0];
  return join qq{\n}, map { $self->_compose_section($_) } $self->section_names_sorted;
}

1;

__END__

=pod

=head1 NAME

String::Sections::Result - Glorified wrapper around a hash representing a parsed String::Sections result

=head1 VERSION

version 0.003003

=head1 METHODS

=head2 C<sections>

    my $sections = $result->sections;
    for my $key( keys %{$sections}) {
        ...
    }

=head2 C<set_current>

    $result->set_current('foo');

=head2 C<has_current>

    if ( $result->has_current ){
    }

=head2 C<section>

    my $ref = $result->section( $name );
    print ${$ref};

=head2 C<section_names>

This contains the names of the sections in the order they were found/inserted.

    my @names = $result->section_names;

=head2 C<section_names_sorted>

=head2 C<has_section>

    if ( $result->has_section($name) ) {
        ...
    }

=head2 C<set_section>

    $result->set_section($name, \$data);

=head2 C<append_data_to_current_section>

    # Unitialise slot
    $result->append_data_to_current_section();
    # Unitialise and/or extend slot
    $result->append_data_to_current_section('somedata');

=head2 C<append_data_to_section>

    # Unitialise slot
    $result->append_data_to_current_section( $name );
    # Unitialise and/or extend slot
    $result->append_data_to_current_section( $name, 'somedata');

=head2 C<shallow_clone>

    my $clone = $result->shallow_clone;

    if ( refaddr $clone->section('foo') == refaddr $result->section('foo') ) {
        print "clone success!"
    }

=head2 C<shallow_merge>

    my $merged = $result->shallow_merge( $other );

    if ( refaddr $merged->section('foo') == refaddr $result->section('foo') ) {
        print "foo copied from orig successfully!"
    }
    if ( refaddr $merged->section('bar') == refaddr $other->section('bar') ) {
        print "bar copied from other successfully!"
    }

=head2 C<to_s>

    my $str = $result->to_s

=head1 ATTRIBUTES

=head2 C<sections>

=head1 PRIVATE ATTRIBUTES

=head2 C<_current>

=head2 C<_section_names>

=head1 PRIVATE METHODS

=head2 C<_current>

    my $current = $result->_current;

=head2 <_section_names>

=head2 C<_compose_section>

    my $str = $result->_compose_section('bar');

=begin MetaPOD::JSON v1.1.0

{
    "namespace":"String::Sections::Result",
    "interface":"class",
    "inherits":"Moo::Object"
}


=end MetaPOD::JSON

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2015 by Kent Fredric <kentnl@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


devel/compile_test_templates.pl
#!/usr/bin/env perl 

use strict;
use warnings;
use utf8::all;

my $project_root;
my $corpus;
my $template_root;
my $output_root;

BEGIN {
  use Path::Tiny;
  use FindBin;
  $project_root  = path($FindBin::Bin)->parent;
  $corpus        = $project_root->child('corpus');
  $template_root = $corpus->child('template');
  $output_root   = $project_root->child('t')->child('02-String-Sections');
}

use Template::Alloy;
use lib $project_root->child('lib')->stringify;

my $targets = [
  {
    template   => 'parse_filehandle.tpl',
    target_dir => $output_root->child('parse_filehandles'),
  },
  {
    template   => 'parse_list.tpl',
    target_dir => $output_root->child('parse_list'),
  }
];
use String::Sections;
use Data::Dump qw(pp);
my $parser = String::Sections->new();

for my $target ( @{$targets} ) {
  *STDOUT->printf( "* Processing target %s into %s\n", $target->{template}, $target->{target_dir} );
  my $t = Template::Alloy->new(
    SYNTAX   => 'tt3',
    filename => $target->{template},
    path     => [ $template_root->stringify ],
  );

  for my $file ( $corpus->child('template_body')->children ) {

    my $filename = $file->basename;
    $filename =~ s/[.][^.]+$/.t/;
    *STDOUT->printf( "  %30s → %-30s\n", $file->basename, $filename );
    my $output_file = $target->{target_dir}->child($filename);
    my $output_fh   = $output_file->openw();

    my $data = $parser->load_filehandle( $file->openr() );
    $t->param( { map { $_, ${ $data->section($_) } } $data->section_names } );
    $t->output( print_to => $output_fh );
    for my $name ( $data->section_names ) {
      delete $t->{_var}->{$name};
    }
    1;
  }
}


Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.034.
use strict;
use warnings;

use Module::Build 0.28;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.4211"
  },
  "configure_requires" => {
    "Module::Build" => "0.4211",
    "perl" => "5.010001"
  },
  "dist_abstract" => "Extract labeled groups of sub-strings from a string.",
  "dist_author" => [
    "Kent Fredric <kentnl\@cpan.org>"
  ],
  "dist_name" => "String-Sections",
  "dist_version" => "0.003003",
  "license" => "perl",
  "module_name" => "String::Sections",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Carp" => 0,
    "Moo" => "1.000008",
    "Scalar::Util" => 0,
    "Types::Standard" => 0,
    "perl" => "5.010001",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => [],
  "test_requires" => {
    "Data::Dump" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "FindBin" => 0,
    "Path::Tiny" => 0,
    "Sub::Exporter::Progressive" => 0,
    "Test::Fatal" => 0,
    "Test::More" => "1.301001_098",
    "lib" => 0,
    "perl" => "5.010001"
  }
);


my %fallback_build_requires = (
  "Data::Dump" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "FindBin" => 0,
  "Module::Build" => "0.4211",
  "Path::Tiny" => 0,
  "Sub::Exporter::Progressive" => 0,
  "Test::Fatal" => 0,
  "Test::More" => "1.301001_098",
  "lib" => 0,
  "perl" => "5.010001"
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;
